packages/itwinui-react/src/core/Menu/MenuItem.tsx

```diff
@@ -139,112 +190,143 @@ export const MenuItem = React.forwardRef((props, forwardedRef) => {
+        onActivate();
         event.preventDefault();
         break;
       }
-      case 'Escape': {
-        // focus might get lost if submenu closes so move it back to parent
-        parent.ref?.current?.focus();
-        break;
-      }
       default:
         break;
     }
   };
 
+  const onMouseEnter = (e: React.MouseEvent<HTMLElement>) => {
+    if (e.target === e.currentTarget) {
+      menuItemRef.current?.focus();
+    }
+  };
+
+  const onFocus = () => {
+    if (parent != null && parentTreeIndex != null) {
+      parent.setActiveIndex(parentTreeIndex);
+    }
+
+    parent.setHasFocusedNodeInSubmenu(true);
+
+    tree?.events.emit('nodeFocused', {
+      nodeId,
+      parentId,
+    } satisfies TreeEvent);
+  };
+
+  const onClick = (e: React.MouseEvent<HTMLElement>) => {
+    onActivate();
+
+    e.preventDefault();
+    e.stopPropagation();
+  };
+
+  const onActivate = () => {
+    !disabled && onClickProp?.(value);
+    setIsSubmenuVisible((prev) => !prev);
+  };
+
   const handlers = {
-    onClick: () => !disabled && onClick?.(value),
+    onClick,
     onKeyDown,
+    onMouseEnter,
+    onFocus,
   };
 
   return (
-    <ListItem
-      as='div'
-      actionable
-      size={size}
-      active={isSelected}
-      disabled={disabled}
-      ref={useMergedRefs(
-        menuItemRef,
-        forwardedRef,
-        subMenuItems.length > 0 ? popover.refs.setReference : null,
-      )}
-      role={role}
-      tabIndex={disabled || role === 'presentation' ? undefined : -1}
-      aria-selected={isSelected}
-      aria-haspopup={subMenuItems.length > 0 ? 'true' : undefined}
-      aria-controls={subMenuItems.length > 0 ? submenuId : undefined}
-      aria-expanded={subMenuItems.length > 0 ? popover.open : undefined}
-      aria-disabled={disabled}
-      {...(subMenuItems.length === 0
-        ? { ...handlers, ...rest }
-        : popover.getReferenceProps({ ...handlers, ...rest }))}
-    >
-      {startIcon && (
-        <ListItem.Icon as='span' aria-hidden>
-          {startIcon}
-        </ListItem.Icon>
-      )}
-      <ListItem.Content>
-        <div>{children}</div>
-        {sublabel && <ListItem.Description>{sublabel}</ListItem.Description>}
-      </ListItem.Content>
-      {!endIcon && subMenuItems.length > 0 && (
-        <ListItem.Icon as='span' aria-hidden>
-          <SvgCaretRightSmall />
-        </ListItem.Icon>
-      )}
-      {endIcon && (
-        <ListItem.Icon as='span' aria-hidden>
-          {endIcon}
-        </ListItem.Icon>
-      )}
+    <>
+      <ListItem
+        as='div'
+        actionable
+        size={size}
+        active={isSelected}
+        disabled={disabled}
+        ref={useMergedRefs(
+          menuItemRef,
+          forwardedRef,
+          subMenuItems.length > 0 ? popover.refs.setReference : null,
+          (el) => {
+            if (parent != null && parentTreeIndex != null) {
+              parent.listItemsRef.current[parentTreeIndex] = el as HTMLElement;
```

Comment:
Can you try to explain in some more detail the new changes in this file since last week? I don't see them explained anywhere, and I'm very confused about how you're handling indexes.

Alternatively, I can reach out privately if you prefer to explain 1:1. After that we can try to improve the PR description and add some code comments.